Source/WebCore/css/CSSCounterStyleRegistry.cpp

```diff
@@ -0,0 +1,156 @@
+            auto decimalCounter = getDecimalCounter();
+            for (const auto counter : countersInChain) {
+                ASSERT(counter);
+                if (!counter)
+                    continue;
+                counter->extendAndResolve(*decimalCounter);
+            }
+            // Recursion return for circular chain.
+            return;
+        }
+        countersInChain.add(&counterStyle);
+
+        auto extendedCounter = counterStyleForName(counterStyle.extendsName());
+        ASSERT(extendedCounter);
+        if (!extendedCounter)
+            return;
+
+        if (extendedCounter->isExtendsSystem() && extendedCounter->isExtendsUnresolved())
+            resolve(*extendedCounter);
+
+        // Recursion return for non-circular chain. Calling resolveExtendsReference() for the extendedCounter might have already resolved this counterStyle if a circle was formed. If it is still unresolved, it should get resolved here.
+        if (counterStyle.isExtendsUnresolved())
+            counterStyle.extendAndResolve(*extendedCounter);
+    };
+
+    resolve(counterStyle);
+}
+
+void CSSCounterStyleRegistry::resolveFallbackReference(CSSCounterStyle& counter)
+{
+    counter.setFallbackReference(counterStyleForName(counter.fallbackName()));
+}
+
+Ref<CSSCounterStyleRegistry> CSSCounterStyleRegistry::create()
+{
+    return adoptRef(*new CSSCounterStyleRegistry());
+}
+
+void CSSCounterStyleRegistry::addCounterStyle(const CSSCounterStyleDescriptors& descriptors)
+{
+    hasUnresolvedReferences = true;
+    m_authorCounterStyles.set(descriptors.m_name, CSSCounterStyle::create(descriptors, false));
+}
+
+RefPtr<CSSCounterStyle> CSSCounterStyleRegistry::getDecimalCounter() const
+{
+    auto& userAgentCounters = userAgentCounterStyles();
+    auto iterator = userAgentCounters.find("decimal"_s);
+    if (iterator != userAgentCounters.end())
+        return iterator->value.get();
+    // user agent counter style should always be populated with a counter named decimal
+    ASSERT_NOT_REACHED();
+    return nullptr;
+}
+
+RefPtr<CSSCounterStyle> CSSCounterStyleRegistry::counterStyleForName(const AtomString& name)
+{
+    resolveReferences();
+    if (name.isEmpty())
+        return getDecimalCounter();
+
+    auto getCounter = [&](const CounterStyleMap& map, const AtomString& counterName) {
+        auto counterIterator = map.find(counterName);
+        return counterIterator != map.end() ? counterIterator->value.get() : nullptr;
+    };
+
+    auto authorCounter = getCounter(m_authorCounterStyles, name);
+    if (authorCounter)
+        return authorCounter;
+    auto userAgentCounter = getCounter(userAgentCounterStyles(), name);
+    if (userAgentCounter)
+        return userAgentCounter;
+
+    return getDecimalCounter();
+}
+
+CounterStyleMap& CSSCounterStyleRegistry::userAgentCounterStyles()
+{
+    // FIXME: counter-style should get pre-defined counters from UA stylesheet rdar://103021161.
```

Comment:
Should we dump the pre-defined ones into our user-agent style sheet instead of writing tons of C++ code to represent them?

Reply:
Yes, this is definitely what we want to do. I've just added decimal manually because it is the last fallback option. Once we get other things in place we can add the counters to UA stylesheet including decimal.